source/registry.d

```diff
@@ -0,0 +1,679 @@
+ * This comes with:
+ * + id: ***20 bits*** (max at `1_048_574`)
+ * + batch: ***12 bits*** (resets at `4_096`)
+ *
+ * Examples:
+ * --------------------
+ * auto registry = Registry();
+ * --------------------
+ */
+alias Registry = BasicRegistry!(uint, 20);
+
+
+@safe pure
+@("registry: batchOf")
+unittest
+{
+	auto registry = new Registry();
+	auto e0 = registry.create;
+	assertEquals(0, registry.batchOf(e0));
+
+	registry.murder(e0);
+	assertEquals(0, registry.batchOf(e0));
+
+	auto e1 = registry.create;
+	assertEquals(1, registry.batchOf(e1));
+	assertEquals(0, registry.batchOf(e0));
+}
+
+
+@safe pure
+@("registry: currentBatchOf")
+unittest
+{
+	auto registry = new Registry();
+	auto e0 = registry.create;
+	assertEquals(0, registry.currentBatchOf(e0));
+
+	registry.murder(e0);
+	assertEquals(1, registry.currentBatchOf(e0));
+}
+
+
+/**
+ * This must be @trusted because std.container : Array is not @safe
+ */
+@trusted pure
+@("registry: entities")
+unittest
+{
+	import std.array : array;
+	auto registry = new Registry();
+
+	auto e0 = registry.create;
+	auto e1 = registry.create;
+	auto e2 = registry.create;
+
+	assertEquals([e0, e1, e2], registry.entities.array);
+
+	registry.murder(e1);
+	const pos1 = registry.idOf(registry.entityNull) | (registry.currentBatchOf(e1) << registry.entityShift);
+	assertEquals([e0, pos1, e2], registry.entities.array);
+
+	registry.murder(e0);
+	const pos0 = registry.idOf(e1) | (registry.currentBatchOf(e0) << registry.entityShift);
+	assertEquals([pos0, pos1, e2], registry.entities.array);
+
+	auto e01 = registry.create;
+	auto e11 = registry.create;
+	assertEquals([e01, e11, e2], registry.entities.array);
+}
+
+
+@safe pure
+@("registry: hasSpawned")
+unittest
+{
+	auto registry = new Registry();
+
+	auto invalid = 0;
+	assertFalse(registry.hasSpawned(invalid));
+
+	auto e0 = registry.create;
+	assertTrue(registry.hasSpawned(e0));
+
+	registry.murder(e0);
+	assertTrue(registry.hasSpawned(e0));
+}
+
+
+@safe pure
+@("registry: idOf")
+unittest
+{
+	auto registry = new Registry();
+
+	auto e0 = registry.create;
+	assertEquals(0, registry.idOf(e0));
+
+	registry.murder(e0);
+	assertEquals(0, registry.idOf(e0));
+
+	auto e01 = registry.create;
+	assertEquals(registry.idOf(e01), registry.idOf(e0));
+	assertNotSame(e0, e01);
+}
+
+
+@safe pure
+@("registry: isValid")
+unittest
+{
+	auto registry = new Registry();
+
+	auto invalid = 0;
+	assertFalse(registry.isValid(invalid));
+
+	auto e0 = registry.create;
+	assertTrue(registry.isValid(e0));
+
+	registry.murder(e0);
+	assertFalse(registry.isValid(e0));
+
+	auto e01 = registry.create;
+	assertTrue(registry.isValid(e01));
+	assertFalse(registry.isValid(e0));
+}
+
+
+@safe pure
+@("registry: murder")
+unittest
+{
+	auto registry = new Registry();
+	const uint invalid = 1;
+	auto exception = expectThrows!EntityException(registry.murder(invalid));
+	assertEquals("Cannot murder entity with invalid id!", exception.msg);
```

Comment:
It's better to implement subclasses of `EntityException` for each error type, rather than having the exception's message being the only thing you can use to distinguish them *(if I were to have an error, how would I know in code what was the specific error?)*.

Reply:
I'm only using the exception twice. Isn't worse having 2 diferent excptions used once than one exception used twice? And isn't the message explicit enough?

Reply:
No, that's not what I meant: I mean when a user is using your API and programming something with it.

I don't think right now this happens, but I assume this project only gets bigger and that this could happen; imagine if I'm calling a function where two errors could happen; how would I handle specific instructions for one or the other?

```d
try {
    auto entity = registry.createEntityWithID(id); // This function can throw one of two errors: either the registry is full, or the ID is invalid (for example, -1)
} catch(EntityException e) {
   // A error just happened; but which one?? I have no way to know other than comparing the exception's message string.
}

Reply:
done